src/services/inlineHints.ts

```diff
@@ -0,0 +1,300 @@
+/* @internal */
+namespace ts.InlineHints {
+
+    const maxHintsLength = 30;
+
+    export function provideInlineHints(context: InlineHintsContext): InlineHint[] {
+        const { file, program, span, cancellationToken, preferences } = context;
+
+        const checker = program.getTypeChecker();
+        const result: InlineHint[] = [];
+        const callExpressionHintableCache = new Map<CallExpression, boolean>();

```

Comment:
Do we expect ES5 environments anymore? If not, we don't necessarily have support for `new Map()`.

One optimization (don't do it yet) might be to cache across program instances. If you scroll your window slightly, you'll end up having to re-populate this every time.

On the other hand, it would be nice if editors knew not to throw away prior inline hints if you've scrolled so the language service won't get asked about them again.

Again, don't do it! It's unclear if this is even a problem yet.